libsecret: enable TPM support

illdefined opened the pull request:
libsecret supports binding the encryption key used for the file backend to a TPM, which offers security benefits over just deriving it from the user’s login password: https://gnome.pages.gitlab.gnome.org/libsecret/libsecret-tpm2.html

This change adds the option `withTpm2Tss` to enable building with TPM support. ~I suggest to enable it by default on x86-64 considering the almost universal prevalence of TPMs on x86-64 desktop systems.~

The option `abrmdSupport` is used to enable support for the user‐space TPM resource manager.  When enabled, `libtss2-tcti-tabrmd.so` and `libtss2-tcti-device.so` are added to the library’s dependencies to enable loading at run‐time through `dlopen()`.

libsecret requires a TPM resource manager, but works fine with the kernel‐space resource manager (`/dev/tpmrm0`). When using an emulated TPM (e.g. libtpms or swtpm), a user‐space resource manager (abrmd) has to be used.

## Things done

I tested this change on nixpkgs master.

The package tests run successfully and I manually tested `secret-tool` using the hardware TPM of my computer (NixOS x86-64) and the kernel‐space resource manager.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

illdefined committed:
libsecret: enable TPM support

gaelj commented:
> I suggest to enable it by default on x86-64 considering the almost universal prevalence of TPMs on x86-64 desktop systems.

IMO that's a bit of a bold clame in general, and especially on Linux where we supposedly have an influx of disgruntled windows 10 users making the switch just because of that.
Will it fail gracefully - or at least, with a helpful error message - on systems that do not have TPM ?

illdefined commented:
> > I suggest to enable it by default on x86-64 considering the almost universal prevalence of TPMs on x86-64 desktop systems.
> 
> IMO that's a bit of a bold clame in general, and especially on Linux where we supposedly have an influx of disgruntled windows 10 users making the switch just because of that. Will it fail gracefully - or at least, with a helpful error message - on systems that do not have TPM ?

libsecret will by default rely on an external D-Bus service (`org.freedesktop.Secret`, as provided by gnome-keyring or KDE Wallet): <https://gitlab.gnome.org/GNOME/libsecret/-/blob/main/libsecret/secret-backend.c?ref_type=heads#L164> 

The file backend can be used instead by setting the environment variable `SECRET_BACKEND` to `file`.

If the TPM is not available (I removed the necessary permissions from the TPM device node in this example), `secret-tool` will issue the following error message:

```
ERROR:tcti:src/tss2-tcti/tcti-device.c:455:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tpmrm0: Permission denied 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:159:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-device.so.0 
ERROR:tcti:src/tss2-tcti/tcti-device.c:455:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tpm0: Permission denied 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:159:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-device.so.0 
ERROR:tcti:src/tss2-tcti/tcti-device.c:455:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tcm0: No such file or directory 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:159:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-device.so.0 
WARNING:tcti:src/util-io/io.c:262:socket_connect() Failed to connect to host 127.0.0.1, port 2321: errno 111: Connection refused 
ERROR:tcti:src/tss2-tcti/tcti-swtpm.c:617:Tss2_Tcti_Swtpm_Init() Cannot connect to swtpm TPM socket 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:159:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-swtpm.so.0 
WARNING:tcti:src/util-io/io.c:262:socket_connect() Failed to connect to host 127.0.0.1, port 2321: errno 111: Connection refused 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:159:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-mssim.so.0 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:273:tctildr_get_default() No standard TCTI could be loaded 
ERROR:tcti:src/tss2-tcti/tctildr.c:477:tctildr_init_context_data() Failed to instantiate TCTI 
secret-tool: Tss2_TctiLdr_Initialize: tcti:IO failure
```

The exact error presentation will of course depend on the application.

gaelj commented:
Thanks for clarifying. That error message is sufficiently informative as far as I'm concerned.

jtojnar commented:
Is not the `file` backend just meant for Flatpak? It looks like it will use a [single path](https://gitlab.gnome.org/GNOME/libsecret/-/blob/ae00f7726c60ff319e3cbd83455c5c80b06bc480/libsecret/secret-file-backend.c#L78-81) so that would conflict if multiple different programs used it. The [docs](https://gitlab.gnome.org/GNOME/libsecret/-/merge_requests/146) also seem to say that it is bound to the secrets portal. That would explain why it is [disabled by default](https://gitlab.gnome.org/GNOME/libsecret/-/blob/ae00f7726c60ff319e3cbd83455c5c80b06bc480/meson_options.txt#L9).

illdefined commented:
> Is not the `file` backend just meant for Flatpak? It looks like it will use a [single path](https://gitlab.gnome.org/GNOME/libsecret/-/blob/ae00f7726c60ff319e3cbd83455c5c80b06bc480/libsecret/secret-file-backend.c#L78-81) so that would conflict if multiple different programs used it. The [docs](https://gitlab.gnome.org/GNOME/libsecret/-/merge_requests/146) also seem to say that it is bound to the secrets portal. That would explain why it is [disabled by default](https://gitlab.gnome.org/GNOME/libsecret/-/blob/ae00f7726c60ff319e3cbd83455c5c80b06bc480/meson_options.txt#L9).

Looking at `secret-file-backend.c`, I believe that the TPM is only used if libsecret detects that it is not sandboxed (where it would rely on the Secret portal) and would otherwise fail: <https://gitlab.gnome.org/GNOME/libsecret/-/blob/ae00f7726c60ff319e3cbd83455c5c80b06bc480/libsecret/secret-file-backend.c#L596>

When writing to the file, libsecret relies on `g_file_replace_contents_bytes_async()` which attempts to modify the file atomically. I am unfamiliar with the semantics of the `GFile` interface, but I suspect that this permits concurrent access by multiple processes.

My personal motivation for this change is that I have a well‐working setup for password management, but no provider for the D-Bus Secrets Service. Some applications however use libsecret to store authentication tokens, and running gnome-keyring or KDE wallet for these would be overkill. Both also lack support for binding secrets to a TPM, which I find particularly useful for impermanent secrets like authentication tokens. I could hack a solution with [secretsd](https://github.com/grawity/secretsd) or wait for [oo7](https://github.com/bilelmoussaoui/oo7) to be mature enough, but why not just use the existing TPM support in libsecret?

vikanezrimaya reviewed (changes requested):
I like the tests, and generally approve of exposing this as an option, but I don't believe it's worth it to enable this by default, even though I'm a fan of the TPM technology.

jtojnar commented on pkgs/by-name/li/libsecret/package.nix:
```diff
@@ -99,25 +115,80 @@ stdenv.mkDerivation rec {
+
+          export TCTI="tabrmd:bus_type=session"
+        ''
+        + lib.optionalString (!abrmdSupport) ''
```
Why is this negated?

illdefined replied:
It is only needed if the `tabrmd` TCTI module has not already been linked during compile time.

`abrmdSupport`:
  - `true`: Test binaries will locate the `tabrmd` module through the `DT_RUNPATH` and `DT_NEEDED` entries in `libsecret-1.so` added at compile‐time. Allow tests to fail if not properly linked.
  - `false`: Test binaries will locate the `tabrmd` module through the `LD_LIBRARY_PATH` environment variable. It is only required during testing because the TPM emulator `swtpm` lacks an integrated resource manager.

jtojnar replied:
Could you please clarify that in the comment? It looks confusing that the abrmd package is used when the corresponding argument is false.

illdefined commented on pkgs/by-name/li/libsecret/package.nix:
```diff
@@ -25,8 +25,15 @@
   libintl,
   dbus,
+  withTpm2Tss ? lib.meta.availableOn stdenv.hostPlatform tpm2-tss && stdenv.hostPlatform.isx86_64,
+  abrmdSupport ? withTpm2Tss,
```
I disabled the TPM‐related options by default and added a `libsecret-tpm` package enabling `withTpm2Tss` so that Hydra runs the tests regularly.

illdefined commented:
Since I have been rebasing, here is a diff with the changes Vika proposed:

```diff
diff --git a/pkgs/by-name/li/libsecret/package.nix b/pkgs/by-name/li/libsecret/package.nix
index fdc902b3a39a..11db576b6e87 100644
--- a/pkgs/by-name/li/libsecret/package.nix
+++ b/pkgs/by-name/li/libsecret/package.nix
@@ -25,8 +25,8 @@
   gjs,
   libintl,
   dbus,
-  withTpm2Tss ? lib.meta.availableOn stdenv.hostPlatform tpm2-tss && stdenv.hostPlatform.isx86_64,
-  abrmdSupport ? withTpm2Tss,
+  withTpm2Tss ? false,
+  abrmdSupport ? false,
   tpm2-tss,
   swtpm,
   tpm2-abrmd,
@@ -213,6 +213,11 @@ stdenv.mkDerivation rec {
     homepage = "https://gitlab.gnome.org/GNOME/libsecret";
     license = lib.licenses.lgpl21Plus;
     mainProgram = "secret-tool";
-    inherit (glib.meta) platforms maintainers;
+    platforms =
+      if withTpm2Tss then
+        lib.intersectLists glib.meta.platforms tpm2-tss.meta.platforms
+      else
+        glib.meta.platforms;
+    inherit (glib.meta) maintainers;
   };
 }
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index d88e3d92b89d..0c75faabf270 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -9758,6 +9758,8 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) ApplicationServices Foundation;
   };
 
+  libsecret-tpm = libsecret.override { withTpm2Tss = true; };
+
   libsigcxx = callPackage ../development/libraries/libsigcxx { };
 
   libsigcxx30 = callPackage ../development/libraries/libsigcxx/3.0.nix { };
```

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9766,6 +9766,8 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) ApplicationServices Foundation;
   };
 
+  libsecret-tpm = libsecret.override { withTpm2Tss = true; };
```
This should be listed in `passthru.tests` so that it does not get lost.

jtojnar commented on pkgs/by-name/li/libsecret/package.nix:
```diff
@@ -93,19 +153,38 @@ stdenv.mkDerivation rec {
+  preConfigure = lib.optionalString abrmdSupport ''
+    # Add dependencies on TCTI modules required for user‐space TPM resource
+    # manager support so that they can be loaded at run time.
+    mesonFlagsArray+=("-Dc_link_args=-Wl,--push-state,--no-as-needed -ltss2-tcti-tabrmd -ltss2-tcti-device -Wl,--pop-state")
```
Could you please also clarify if these meant for `dlopen`? Or why does the Meson not add them to NEEDED?

illdefined replied:
They are loaded at run time by tpm2-tss through `dlopen()`.

illdefined commented:
I removed dependency on swtpm for TPM emulation in favour of libtpms, which is included in tpm2-tss by default and requires no setup.

I also added a little code to wait for `tpm2-abrmd` to become available on the session bus before initiating the tests, to avoid spurious test failures.
